core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala

```diff
@@ -56,12 +58,30 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
 
   /**
    * Acquire N bytes of memory to cache the given block, evicting existing ones if necessary.
- *
+   *
    * @return whether all N bytes were successfully granted.
    */
-  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = lock.synchronized {
-    val numBytesToFree = math.max(0, numBytes - memoryFree)
-    acquireMemory(blockId, numBytes, numBytesToFree)
+  def acquireMemory(blockId: BlockId, numBytes: Long): Boolean = {
+    // First, attempt to acquire as much memory as we can without performing any eviction
+    val bytesAcquiredWithoutEviction = lock.synchronized {
+      val bytesAcquired = math.min(numBytes, memoryFree)
+      _memoryUsed += bytesAcquired
+      bytesAcquired
+    }
+    // Acquire the remainder of the memory by performing eviction
+    try {
+      val bytesToAcquireViaEviction = numBytes - bytesAcquiredWithoutEviction
```

Comment:
can you add some asserts here to make sure this is non-negative? Also I would do all of this only if this is > 0.